File: 1_reading_mnist_labels/c/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int magic_number;
    int num_items;
    unsigned char *labels;
} FileResponse;

FileResponse read_file(const char* file_name) {
    FileResponse response;
    FILE* file = fopen(file_name, "rb");

    if (file == NULL) {
        printf("Error opening file.\n");
        exit(1);
    }

    // Read magic number
    fread(&response.magic_number, sizeof(int), 1, file);
    response.magic_number = htonl(response.magic_number); // Convert to big-endian if necessary

    // Read number of items
    fread(&response.num_items, sizeof(int), 1, file);
    response.num_items = htonl(response.num_items); // Convert to big-endian if necessary

    // Allocate memory for the labels
    response.labels = (unsigned char*)malloc(response.num_items * sizeof(unsigned char));
    if (response.labels == NULL) {
        printf("Memory allocation error.\n");
        fclose(file);
        exit(1);
    }

    // Read labels
    fread(response.labels, sizeof(unsigned char), response.num_items, file);

    fclose(file);
    return response;
}

void free_file_response(FileResponse *response) {
    free(response->labels);
}

void process_file(const char* file_name) {
    FileResponse response = read_file(file_name);

    printf("File: %s\n", file_name);
    printf("Magic Number: %d\n", response.magic_number);
    printf("Number of Items: %d\n", response.num_items);

    printf("Labels:\n");
    for (int i = 0; i < response.num_items; i++) {
        printf("[%d] %u\n", i, response.labels[i]);
    }

    printf("\n");

    free_file_response(&response);
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: %s file1 file2 file3 ...\n", argv[0]);
        return 1;
    }

    for (int i = 1; i < argc; i++) {
        process_file(argv[i]);
    }

    return 0;
}

File: 1_reading_mnist_labels/go/src/main.go
package main

import (
	"fmt"
	"log"
	"os"
)

// FileResponse represents the data structure to hold the contents of the file.
type FileResponse struct {
	MagicNumber int    // The magic number read from the file header.
	NumItems    int    // The number of items read from the file header.
	Labels      []byte // An array to store the label values.
}

// readFile reads the file with the given fileName and returns a pointer to FileResponse.
func readFile(fileName string) (*FileResponse, error) {
	file, err := os.Open(fileName) // Open the file in read mode.
	if err != nil {
		return nil, err // Return error if file cannot be opened.
	}
	defer file.Close() // Defer closing the file until the function returns.

	response := &FileResponse{}
	headerBytes := make([]byte, 8) // A buffer to read the first 8 bytes (magic number and num_items).

	// Read the first 8 bytes (magic number and num_items) from the file header.
	_, err = file.Read(headerBytes)
	if err != nil {
		return nil, err // Return error if reading the header fails.
	}

	// Convert the first 4 bytes to the magic number (big-endian encoding).
	response.MagicNumber = int(headerBytes[0])<<24 | int(headerBytes[1])<<16 | int(headerBytes[2])<<8 | int(headerBytes[3])

	// Convert the next 4 bytes to the number of items (big-endian encoding).
	response.NumItems = int(headerBytes[4])<<24 | int(headerBytes[5])<<16 | int(headerBytes[6])<<8 | int(headerBytes[7])

	// Read the label bytes from the file.
	response.Labels = make([]byte, response.NumItems)
	_, err = file.Read(response.Labels)
	if err != nil {
		return nil, err // Return error if reading the labels fails.
	}

	return response, nil // Return the FileResponse pointer with data.
}

// processFile reads the file with the given fileName and prints its contents.
func processFile(fileName string) {
	response, err := readFile(fileName) // Read the file contents.
	if err != nil {
		log.Fatalf("Error reading file %s: %v", fileName, err) // Handle error if reading fails.
	}

	fmt.Printf("File: %s\n", fileName)
	fmt.Printf("Magic Number: %d\n", response.MagicNumber)
	fmt.Printf("Number of Items: %d\n", response.NumItems)

	fmt.Println("Labels:")
	for i, label := range response.Labels {
		fmt.Printf("[%d] %d\n", i, label) // Print each label with its index.
	}
	fmt.Println()
}

func main() {
	if len(os.Args) < 2 {
		fmt.Printf("Usage: %s file1 file2 file3 ...\n", os.Args[0])
		os.Exit(1)
	}

	for _, fileName := range os.Args[1:] {
		processFile(fileName) // Process each file specified in the command-line arguments.
	}
}